File: backend/app/models/task.rb
class Task < ApplicationRecord
  has_many :items

  def self.allTasksSerialized
    tasks = self.all
    tasksSerialized = {}
    tasks.each do |task|
      jsonItems = {} 
      task.items.each do |item|
        jsonItems[item.id] = {
          id: item.id,
          text: item.text,
          complete: item.complete
        }
      end
      jsonTask = {
        title: task.title, 
        description: task.description,
        items: jsonItems
      }
      tasksSerialized[task.id] = jsonTask
    end
    tasksSerialized
  end
end

File: frontend/src/Task.js
class Task {
  static all = {}

  constructor(json) {
    this.id = json.id;
    this.title = json.title;
    this.description = json.description;
    this.items = Item.createItemsFromJson(json.items, this);
    Task.all[this.id] = this
  }

  static get all() {
    return all
  }

  static createTasksFromJson(json) {
    for(const [key, value] of Object.entries(json)) {
      new Task(value);
    }
  }

  static create(title, description) {
    fetch('http://127.0.0.1:3000/tasks', {
      method: 'POST',
      headers: {
        "Content-Type": "application/json",
        "Accept": "application/json"
      },
      body: JSON.stringify({
        title: title,
        description: description
      })
    })
    .then(res => res.json())
    .then(json => {
      const task = new Task({id: json.id, title: title, description: description, items: {}});
      renderTask(task);
    })
    .catch(err => console.error(err));
  }

  delete() {
    fetch(`http://127.0.0.1:3000/tasks/${this.id}`, {
      method: 'DELETE'
    })
    .catch(err => alert(err));
  }
}